File: Binary_tree/二叉树_纲领/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# 遇到子树问题，首先想到的是给函数设置返回值，然后在后序位置做文章。
class Solution(object):
    maxdiameter = 0

    def diameterOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def traverse(root):
            if root == None:
                return 0
            left_depth = traverse(root.left)
            right_depth = traverse(root.right)
            cur_node_diameter = left_depth+right_depth
            self.maxdiameter = max(self.maxdiameter, cur_node_diameter)
            return max(left_depth, right_depth)+1
        traverse(root)
        return self.maxdiameter

File: Binary_tree/二叉树_纲领/diameterOfBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    int maxdiameter = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        traverse(root);
        return maxdiameter;
    }

    private int traverse(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int leftDepth = traverse(root.left);
        int rightDepth = traverse(root.right);
        // Utilize the post order traversal to calculate the diameter

        int curNodeDiameter = leftDepth + rightDepth;
        maxdiameter = Math.max(maxdiameter, curNodeDiameter);
        return Math.max(leftDepth, rightDepth) + 1;
    }
}